crates/red_knot_python_semantic/src/types.rs

```diff
@@ -3083,6 +3083,15 @@ pub enum KnownConstraintFunction {
     IsSubclass,
 }
 
+impl KnownConstraintFunction {
+    fn apply_constraint(self, class: Class) -> Type {
```

Comment:
Nit: this method doesn't actually "apply" a constraint, it just returns a type. Maybe `constraint_ty`?

I also wonder if this might be even clearer if it accepts an `argument: Type` instead of `class: Class` and encapsulates all of the logic that is currently in `generate_classinfo_constraint` (that is, match tuples and class literals, recursively build a union for tuples)? Don't feel strongly, though.